### PR TITLE
New project and architecture for integration tests & cypress

### PR DESCRIPTION
**Pull request by dborgen2**
## What’s changing?

- New dedicated project for Query Tool Cypress tests that spins up two version of the website (localhost:5001 for mock EB user and localhost:5002 for mock National user). Tests that used to be locally run only can now be run in CICD by calling Azure Functions via reflection.
- You can also run the Cypress project for normal development, as it only requires access to the database. You specify the database connection strings in the launchSettings.json file of the new project. This should take some time away from having to run the IAC to do basic development. However you will still need to have your databases up and running.
- The underlying architecture spins up a mock server and maps the mock server port to the generated HttpClient. We can also use this architecture for doing backend-only integration tests with mocks.
- Additional Cypress tests should be written that utilize this new architecture. I added one, and moved some from the local folder to the CICD folder, but did not consider other new test cases.
- If there are no issues with this architecture, a small follow-up story to do the same for Dashboard should be considered.

## Why?

Closes NAC-2126

## This PR has:

- [x] **JIRA ticket(s)** (_Must be listed in PR request and branch name should follow convention NAC-123-description_)<description>

- [ ] **Security Related Changes** (_Security Tab questions have been answered on JIRA ticket(s)_)

- [x] **Commented source code** (_required on public classes and methods, and elsewhere as appropriate_)

- [ ] **Documented API changes** (_required on public classes and methods, and elsewhere as appropriate_)
    - [ ] External API documentation - OpenApi docs and Generated Docs (e.g. https://github.com/18F/piipan/tree/dev/docs/openapi) 
    - [ ] Internal API documentation (e.g. https://github.com/18F/piipan/tree/dev/metrics/docs/openapi)   

- [ ] **Developer documentation** (_for new build/test/API changes or complex portions of the system_)

- [x] **Automated unit tests** (_to maintain or increase level of code coverage_)

- [ ] **Changes to IAC** (_add any deployment steps that require manual intervention to the draft release notes_)
    - [ ] Have IaC scripts run successfully against an existing deployed environment?
    - [ ] Have IaC scripts run successfully against a newly deployed environment?
    - [ ] Have any new bash scripts been given the appropriate permissions? (e.g. git update-index --chmod=+x $script)
    - [ ] Do the IaC changes require deleting existing resources?

## Anything else?

_Add other details that are helpful for review here_

## Code Review Checklist:

- [ ] Automated checks all pass? (code builds successfully, tests pass, test coverage meets/exceeds threshold, etc.) 
- [ ] Have all requirements and Acceptance Criteria been met?
- **If there are code changes**
    - [ ] Is there no unnecessary complexity? Code is readable, maintainable, etc.
    - [ ] Do the changes follow SOLID principles?
    - [ ] Are the changes robust? Concurrency considered? Performance? Security (e.g. no sql injections)?
    - [ ] Are newly added dependencies needed and appropriate? Are dependency files (e.g. packages.json & packages.lock.json) updated?
    - [ ] For any new projects/solutions, have they been added to code coverage files?
    - [ ] Public classes & methods are documented    
- **If there are API changes**
    - [ ] Does API make sense? Useful and concise?
    - [ ] Are existing API interfaces/contracts intact? i.e. no breaking changes
    - [ ] Are API tests included?
    - [ ] Are API responses appropriate and documented? (200s, 201, 404s, etc.)
    - [ ] Are internals kept private and not leaked into the API?
    - [ ] External API documentation - OpenApi docs and Generated Docs 
    - [ ] Internal API documentation    
- **If there are IAC changes**
    - [ ] IAC Scripts are readable & maintainable
    - [ ] IAC Scripts are documented
    - [ ] IAC scripts pass required checks
    - [ ] IAC scripts were run successfully by author
- **General Logging Questions**
    - [ ] Do new capabilities (or changed capabilities) include logging?
    - [ ] Is PII not logged?
- **General Error Handling Questions**
    - [ ] Has exception handling been added for new capabilities (or changed capabilities)?
    - [ ] Are errors/exceptions properly communicated to end-users? Were there requirements for user facing error messages?
    - [ ] Are errors/exceptions logged?
    - [ ] Is PII kept out of error/exception logging?
- **If there are Developer Documentation Changes**
    - [ ] General Project documentation (under piipan\docs. e.g. iac.md, architecture.md, etc.)
    - [ ] Github documentation additions/changes (Readmes)
- **General automated test questions**
    - [ ] End-to-End Tests provided (if code implements new user story)
    - [ ] Are tests sufficient?
    - [ ] Corner cases considered?
    - [ ] Are performance tests needed?

### CODE REVIEW

File: match/tests/Piipan.Match.Func.ResolutionApi.IntegrationTests/GetMatchApiIntegrationTests.cs [hunk 1]
```diff
@@ -2,67 +2,23 @@
...
-using Piipan.Shared.Database;
-using Piipan.States.Core.DataAccessObjects;
-using Piipan.States.Core.Service;
 using Xunit;
 
 namespace Piipan.Match.Func.ResolutionApi.IntegrationTests
 {
     [Collection("MatchResolutionApiTests")]
-    public class GetMatchApiIntegrationTests : DbFixture
+    public class GetMatchApiIntegrationTests : BaseMatchResIntegrationTests
```
Comment (bgfoshay) [inline match/tests/Piipan.Match.Func.ResolutionApi.IntegrationTests/GetMatchApiIntegrationTests.cs:16]: Praise: The encapsulation of all of the service config and api creation in the Base test classes looks great. Thanks for getting rid of all of the duplication. The tests look a lot more straight-forward now.

File: match/tests/Piipan.Match.Func.ResolutionApi.IntegrationTests/packages.lock.json [hunk 2]
```diff
@@ -546,6 +546,39 @@
           "System.Text.Encodings.Web": "4.5.0"
         }
       },
+      "Microsoft.Azure.Cosmos.Table": {
```
Comment (bgfoshay) [inline match/tests/Piipan.Match.Func.ResolutionApi.IntegrationTests/packages.lock.json:549]: Question/Nitpick: Are Cosmos and OData references being used or were they accidentally pulled in?

File: query-tool/tests/Piipan.QueryTool.Cypress.Tests/create-hash_type.sql [hunk 3]
```diff
@@ -0,0 +1,6 @@
+DO $$ BEGIN
```
Comment (bgfoshay) [inline query-tool/tests/Piipan.QueryTool.Cypress.Tests/create-hash_type.sql:1]: Question: Is this file needed? It appears to be a duplicate of the migration script in collaboration database scripts?

File: query-tool/tests/Piipan.QueryTool.Cypress.Tests/insert-state-info.sql [hunk 4]
```diff
@@ -0,0 +1,38 @@
+BEGIN;
```
Comment (bgfoshay) [inline query-tool/tests/Piipan.QueryTool.Cypress.Tests/insert-state-info.sql:1]: Question: Same question as above regarding these sql scripts and whether it's necessary to duplicate them?

File: query-tool/tests/Piipan.QueryTool.Cypress.Tests/match-record.sql [hunk 5]
```diff
@@ -0,0 +1,53 @@
+BEGIN;
```
Comment (bgfoshay) [inline query-tool/tests/Piipan.QueryTool.Cypress.Tests/match-record.sql:1]: Question: Same question as above regarding these sql scripts and whether it's necessary to duplicate them?

File: query-tool/tests/Piipan.QueryTool.Cypress.Tests/per-state.sql [hunk 6]
```diff
@@ -0,0 +1,50 @@
+BEGIN;
```
Comment (bgfoshay) [inline query-tool/tests/Piipan.QueryTool.Cypress.Tests/per-state.sql:1]: Question: Same question as above regarding these sql scripts and whether it's necessary to duplicate them?

File: query-tool/tests/Piipan.QueryTool.Tests/Components/MatchDetail/InitialActionTests.razor [hunk 7]
```diff
@@ -72,7 +72,7 @@
                 panel.</p>
                 <div diff:ignore> //ignoreing Vulnerable Invalid Component
                 </div>
-                <div id="initial-action-section" class="">
+                <div id="initial-action-section" class="" aria-disabled="false">
```
Comment (bgfoshay) [inline query-tool/tests/Piipan.QueryTool.Tests/Components/MatchDetail/InitialActionTests.razor:75]: Question: Why do these need to be enabled? To make it easier to configure/modify them for tests?

File: shared/tests/Piipan.Shared.Tests/TestFixtures/PiipanAzureFunctionServer.cs [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+
+        /// <summary>
+        /// Creates a Azure Function mock server that runs at localhost:port
+        /// All APIs passed in as params are available to call from the client.
+        /// </summary>
+        /// <typeparam name="TTest">The IIntegration test that we can call to setup the environment variables/services</typeparam>
+        /// <param name="port">The port that this mock server runs on</param>
+        /// <param name="apis">All of the APIs that need to be accounted for</param>
+        /// <returns>The Azure Function mock server</returns>
+        public static PiipanAzureFunctionServer CreateWithPort<TTest>(string port, params Type[] apis) where TTest : IIntegrationTest, new()
```
Comment (bgfoshay) [inline shared/tests/Piipan.Shared.Tests/TestFixtures/PiipanAzureFunctionServer.cs:35]: Question: Does this method need to be generic? It looks like it'll work just fine, but from a code complexity standpoint, I get more nervous the deeper the level of generic method calling goes. In this case, it looks like it's newing up an instance of TTest and then calling SetupServices()? So it may not be a big deal. Just thought I'd ask.

File: states/src/Piipan.States/Piipan.States.Api/IStatesApi.cs [hunk 9]
```diff
@@ -6,6 +6,6 @@ public interface IStatesApi
     {
         Task<StatesInfoResponse> GetStates();
 
-        Task<StatesInfoResponse> UpsertState(StateInfoRequest stateInfoRequest);
+        Task<string> UpsertState(StateInfoRequest stateInfoRequest);
```
Comment (bgfoshay) [inline states/src/Piipan.States/Piipan.States.Api/IStatesApi.cs:9]: Question: Just curious as to what drove the change from strongly-typed to string?

Comment (dborgen2) [reply states/src/Piipan.States/Piipan.States.Api/IStatesApi.cs:9]: This was actually a bug that I found while testing. UpsertState ended up just returning a string, which would cause an exception when calling it from the StatesClient. I'm guessing the testing on it was done using the API itself, not the client.

File: shared/tests/Piipan.Shared.Tests/TestFixtures/PiipanAzureFunctionServer.cs (hunk 8 above)
Comment (dborgen2) [reply shared/tests/Piipan.Shared.Tests/TestFixtures/PiipanAzureFunctionServer.cs:35]: Yea, the altnerative would be to call CreateWithPort, passing in the instance with a type of TTest instead of creating it here. It's needed in some capacity to call SetupServices(). If you think it would be cleaner to pass it in as a parameter, that's fine, I can do that no problem.

File: query-tool/tests/Piipan.QueryTool.Tests/Components/MatchDetail/InitialActionTests.razor (hunk 7 above)
Comment (dborgen2) [reply query-tool/tests/Piipan.QueryTool.Tests/Components/MatchDetail/InitialActionTests.razor:75]: Actually one of the cypress tests that I added found an accessibility issue. When this section gets disabled, a class gets added but aria-disabled was never set, so someone using a screen reader wouldn't know the area is disabled.

File: query-tool/tests/Piipan.QueryTool.Cypress.Tests/Startup.cs [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+        protected override void UseMockFile(IServiceCollection services)
+        {
+            var mockFile = $"{_env.ContentRootPath}/{_mockFile}";
+            services.UseJsonFileToMockEasyAuth(mockFile);
+        }
+
+        protected override void RegisterMatchClientServices(IServiceCollection services)
+        {
+            var clientFactory = services.FirstOrDefault(n => n.ServiceType == typeof(IHttpClientFactory));
+            services.Remove(clientFactory);
```
[the remark this answers is not in the document]
Comment (dborgen2) [reply query-tool/tests/Piipan.QueryTool.Cypress.Tests/Startup.cs:44]: Yes, that's correct. Removing it and replacing it with this new HttpClientClientFactory that gets added during RegisterApi

File: match/tests/Piipan.Match.Func.ResolutionApi.IntegrationTests/packages.lock.json (hunk 2 above)
Comment (dborgen2) [reply match/tests/Piipan.Match.Func.ResolutionApi.IntegrationTests/packages.lock.json:549]: I think this is an accident, as I was doing some researching regarding OData and probably forgot to remove the references. Let me double check

File: query-tool/tests/Piipan.QueryTool.Cypress.Tests/Startup.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (bgfoshay) [reply query-tool/tests/Piipan.QueryTool.Cypress.Tests/Startup.cs:44]: 👍

File: query-tool/tests/Piipan.QueryTool.Tests/Components/MatchDetail/InitialActionTests.razor (hunk 7 above)
Comment (bgfoshay) [reply query-tool/tests/Piipan.QueryTool.Tests/Components/MatchDetail/InitialActionTests.razor:75]: ok, gotcha. In that case, nice find. Tests are already paying off 🙂

File: match/tests/Piipan.Match.Func.ResolutionApi.IntegrationTests/packages.lock.json (hunk 2 above)
Comment (dborgen2) [reply match/tests/Piipan.Match.Func.ResolutionApi.IntegrationTests/packages.lock.json:549]: Actually no, it looks like these already existed, but were now added because I added a reference to Piipan.Shared.Tests.csproj. I don't think there is anything to change here.

Comment (bgfoshay) [reply match/tests/Piipan.Match.Func.ResolutionApi.IntegrationTests/packages.lock.json:549]: Interesting. Ok, sounds good. We can look at this in more detail later.

File: query-tool/tests/Piipan.QueryTool.Cypress.Tests/per-state.sql (hunk 6 above)
Comment (dborgen2) [reply query-tool/tests/Piipan.QueryTool.Cypress.Tests/per-state.sql]: Fixed

File: query-tool/tests/Piipan.QueryTool.Cypress.Tests/match-record.sql (hunk 5 above)
Comment (dborgen2) [reply query-tool/tests/Piipan.QueryTool.Cypress.Tests/match-record.sql]: Fixed

File: query-tool/tests/Piipan.QueryTool.Cypress.Tests/insert-state-info.sql (hunk 4 above)
Comment (dborgen2) [reply query-tool/tests/Piipan.QueryTool.Cypress.Tests/insert-state-info.sql]: Fixed

File: query-tool/tests/Piipan.QueryTool.Cypress.Tests/create-hash_type.sql (hunk 3 above)
Comment (dborgen2) [reply query-tool/tests/Piipan.QueryTool.Cypress.Tests/create-hash_type.sql]: Fixed

**Review by bgfoshay — APPROVED** [review]
LGTM. This is some great work. These changes are going to be really useful for both running cypress tests in CICD and for the developers in their local testing.
